File: backend/src/main/kotlin/app/meals/domain/UnitConversion.kt
package app.meals.domain

/**
 * Unit families for shopping-list aggregation. Quantities in a family are converted to a base unit
 * (ml for volume, g for weight), summed, then formatted with [bestDisplayUnit].
 */
enum class UnitFamily {
    VOLUME,
    WEIGHT,
}

/**
 * @param family volume or weight family
 * @param toBase multiplier from this unit to base (ml or g)
 * @param canonical short label for the unit (for documentation; display uses [bestDisplayUnit])
 */
data class KnownUnit(
    val family: UnitFamily,
    val toBase: Double,
    val canonical: String,
)

private val knownByAlias: Map<String, KnownUnit> = buildMap {
    fun putAllAliases(vararg aliases: String, unit: KnownUnit) {
        for (a in aliases) {
            put(a.lowercase(), unit)
        }
    }

    val g = KnownUnit(UnitFamily.WEIGHT, 1.0, "g")
    putAllAliases("g", "gram", "grams", unit = g)

    val hg = KnownUnit(UnitFamily.WEIGHT, 100.0, "hg")
    putAllAliases("hg", "hektogram", "hectogram", "hektograms", "hectograms", unit = hg)

    val kg = KnownUnit(UnitFamily.WEIGHT, 1000.0, "kg")
    putAllAliases("kg", "kilo", "kilogram", "kilograms", unit = kg)

    val tsp = KnownUnit(UnitFamily.VOLUME, 5.0, "tsp")
    putAllAliases(
        "tsp",
        "ts",
        "teskje",
        "teaspoon",
        "teaspoons",
        unit = tsp,
    )

    val tbsp = KnownUnit(UnitFamily.VOLUME, 15.0, "tbsp")
    putAllAliases(
        "tbsp",
        "ss",
        "spiseskje",
        "tablespoon",
        "tablespoons",
        unit = tbsp,
    )

    val ml = KnownUnit(UnitFamily.VOLUME, 1.0, "ml")
    putAllAliases("ml", "milliliter", "milliliters", "millilitre", "millilitres", unit = ml)

    val cl = KnownUnit(UnitFamily.VOLUME, 10.0, "cl")
    putAllAliases("cl", "centiliter", "centiliters", "centilitre", "centilitres", unit = cl)

    val dl = KnownUnit(UnitFamily.VOLUME, 100.0, "dl")
    putAllAliases("dl", "desiliter", "deciliter", "deciliters", "decilitre", "decilitres", unit = dl)

    val l = KnownUnit(UnitFamily.VOLUME, 1000.0, "l")
    putAllAliases("l", "liter", "liters", "litre", "litres", unit = l)
}

/**
 * Returns [KnownUnit] when [unit] is exactly a known alias (trimmed, case-insensitive).
 * Compound strings like "g can" return null.
 */
fun normalizeUnit(unit: String): KnownUnit? {
    val key = unit.trim().lowercase()
    if (key.isEmpty()) return null
    return knownByAlias[key]
}

/**
 * Picks the largest unit where the displayed value is >= 1.
 * [baseValue] is in ml (volume) or g (weight).
 */
fun bestDisplayUnit(baseValue: Double, family: UnitFamily): Pair<Double, String> =
    when (family) {
        UnitFamily.WEIGHT ->
            if (baseValue >= 1000.0) {
                baseValue / 1000.0 to "kg"
            } else {
                baseValue to "g"
            }
        UnitFamily.VOLUME ->
            when {
                baseValue >= 1000.0 -> baseValue / 1000.0 to "l"
                baseValue >= 100.0 -> baseValue / 100.0 to "dl"
                else -> baseValue to "ml"
            }
    }

File: backend/src/domain/unitConversion.ts
// Unit families for shopping-list aggregation. Quantities in a family are
// converted to a base unit (ml for volume, g for weight), summed, then formatted
// with bestDisplayUnit. Ported from the Kotlin UnitConversion.

export type UnitFamily = "volume" | "weight";

export interface KnownUnit {
  family: UnitFamily;
  /** Multiplier from this unit to the base (ml or g). */
  toBase: number;
  canonical: string;
}

const knownByAlias = new Map<string, KnownUnit>();

function register(unit: KnownUnit, ...aliases: string[]): void {
  for (const alias of aliases) knownByAlias.set(alias.toLowerCase(), unit);
}

register({ family: "weight", toBase: 1, canonical: "g" }, "g", "gram", "grams");
register(
  { family: "weight", toBase: 100, canonical: "hg" },
  "hg",
  "hektogram",
  "hectogram",
  "hektograms",
  "hectograms",
);
register({ family: "weight", toBase: 1000, canonical: "kg" }, "kg", "kilo", "kilogram", "kilograms");
register({ family: "volume", toBase: 5, canonical: "tsp" }, "tsp", "ts", "teskje", "teaspoon", "teaspoons");
register(
  { family: "volume", toBase: 15, canonical: "tbsp" },
  "tbsp",
  "ss",
  "spiseskje",
  "tablespoon",
  "tablespoons",
);
register(
  { family: "volume", toBase: 1, canonical: "ml" },
  "ml",
  "milliliter",
  "milliliters",
  "millilitre",
  "millilitres",
);
register(
  { family: "volume", toBase: 10, canonical: "cl" },
  "cl",
  "centiliter",
  "centiliters",
  "centilitre",
  "centilitres",
);
register(
  { family: "volume", toBase: 100, canonical: "dl" },
  "dl",
  "desiliter",
  "deciliter",
  "deciliters",
  "decilitre",
  "decilitres",
);
register({ family: "volume", toBase: 1000, canonical: "l" }, "l", "liter", "liters", "litre", "litres");

/**
 * Returns a KnownUnit when `unit` is exactly a known alias (trimmed,
 * case-insensitive). Compound strings like "g can" return null.
 */
export function normalizeUnit(unit: string): KnownUnit | null {
  const key = unit.trim().toLowerCase();
  if (!key) return null;
  return knownByAlias.get(key) ?? null;
}

/** Pick the largest unit where the displayed value is >= 1. */
export function bestDisplayUnit(baseValue: number, family: UnitFamily): [number, string] {
  if (family === "weight") {
    return baseValue >= 1000 ? [baseValue / 1000, "kg"] : [baseValue, "g"];
  }
  if (baseValue >= 1000) return [baseValue / 1000, "l"];
  if (baseValue >= 100) return [baseValue / 100, "dl"];
  return [baseValue, "ml"];
}
